Userland/Applications/SoundPlayer/BarsVisualizationWidget.cpp

```diff
@@ -0,0 +1,151 @@
+        groups[i] /= max * freq_bin / (m_adjust_frequencies ? (clamp(pow(M_E, (double)i / group_count * 3.) - 1.75, 1., 15.)) : 1.);
+
+    const int horizontal_margin = 30;
+    const int top_vertical_margin = 15;
+    const int pixels_inbetween_groups = frame_inner_rect().width() > 350 ? 5 : 2;
+    int pixel_per_group_width = (frame_inner_rect().width() - horizontal_margin * 2 - pixels_inbetween_groups * (group_count - 1)) / group_count;
+    int max_height = frame_inner_rect().height() - top_vertical_margin;
+    int current_xpos = horizontal_margin;
+    for (int g = 0; g < group_count; g++) {
+        m_gfx_falling_bars[g] = AK::min(clamp(max_height - (int)(groups[g] * max_height * 0.8), 0, max_height), m_gfx_falling_bars[g]);
+        painter.fill_rect(Gfx::Rect(current_xpos, max_height - (int)(groups[g] * max_height * 0.8), pixel_per_group_width, (int)(groups[g] * max_height * 0.8)), Gfx::Color::from_rgb(0x95d437));
+        painter.fill_rect(Gfx::Rect(current_xpos, m_gfx_falling_bars[g], pixel_per_group_width, 2), Gfx::Color::White);
+        current_xpos += pixel_per_group_width + pixels_inbetween_groups;
+        m_gfx_falling_bars[g] += 3;
+    }
+
+    m_is_using_last = false;
+}
+
+BarsVisualizationWidget::~BarsVisualizationWidget()
+{
+}
+
+BarsVisualizationWidget::BarsVisualizationWidget()
+    : m_last_id(-1)
+    , m_is_using_last(false)
+    , m_adjust_frequencies(false)
+{
+    m_context_menu = GUI::Menu::construct();
+    m_context_menu->add_action(GUI::Action::create_checkable("Adjust frequency energy (for aesthetics)", [&](GUI::Action& action) {
+        m_adjust_frequencies = action.is_checked();
+    }));
+}
+
+void BarsVisualizationWidget::set_buffer(RefPtr<Audio::Buffer> buffer, int samples_to_use)
+{
+    if (m_is_using_last)
+        return;
+    m_is_using_last = true;
+    // should probably do a check to see if the buffer contains a meaningful number of samples
+    m_sample_count = round_previous_power_of_2(samples_to_use);
+    m_sample_buffer.ensure_capacity(m_sample_count);
+    m_sample_buffer.resize(m_sample_count);
+    for (int i = 0; i < m_sample_count; i++) {
+        m_sample_buffer.data()[i] = buffer->samples()[i].left;
+    }
+
+    update();
+}
+
+void BarsVisualizationWidget::set_buffer(RefPtr<Audio::Buffer> buffer)
+{
+    if (buffer.is_null())
+        return;
+    if (m_last_id == buffer->id())
+        return;
+    set_buffer(buffer, buffer->sample_count());
+}
+
+void BarsVisualizationWidget::mousedown_event(GUI::MouseEvent& event)
+{
+    Widget::mousedown_event(event);
+    if (event.button() == GUI::Right) {
+        m_context_menu->popup(event.position());
+    }
+}
+
+void BarsVisualizationWidget::set_samplerate(int samplerate)
+{
+    m_samplerate = samplerate;
+}
+
+// black magic from Hacker's delight
+u32 round_previous_power_of_2(u32 x)
+{
+    x = x | (x >> 1);
+    x = x | (x >> 2);
+    x = x | (x >> 4);
+    x = x | (x >> 8);
+    x = x | (x >> 16);
+    return x - (x >> 1);
+}
```

Comment:
How about making this either a local method for this .cpp file, or moving it into AudioAlgorithms.h ?

Reply:
Yeah it didn't really make sense for it to be a member function. Made it a local func.